presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowPageSource.java

```diff
@@ -0,0 +1,163 @@
+    private int currentPosition;
+    private Optional<VectorSchemaRoot> vectorSchemaRoot = Optional.empty();
+    private ArrowFlightClient flightClient;
+    private FlightStream flightStream;
+
+    public ArrowPageSource(ArrowSplit split, List<ArrowColumnHandle> columnHandles, ArrowFlightClientHandler clientHandler,
+            ConnectorSession connectorSession)
+    {
+        this.columnHandles = columnHandles;
+        this.split = split;
+        getFlightStream(clientHandler, split.getTicket(), connectorSession);
+    }
+
+    private void getFlightStream(ArrowFlightClientHandler clientHandler, byte[] ticket, ConnectorSession connectorSession)
+    {
+        try {
+            Optional<String> uri = (split.getLocationUrls().isEmpty()) ?
+                    Optional.empty() : Optional.of(split.getLocationUrls().get(0));
+            flightClient = clientHandler.getClient(uri);
+            flightStream = flightClient.getFlightClient().getStream(new Ticket(ticket), clientHandler.getCallOptions(connectorSession));
+        }
+        catch (FlightRuntimeException e) {
+            throw new ArrowException(ARROW_FLIGHT_ERROR, e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getCompletedPositions()
+    {
+        return currentPosition;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return completed;
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (vectorSchemaRoot.isPresent()) {
+            vectorSchemaRoot.get().close();
+            vectorSchemaRoot = Optional.empty();
+        }
+
+        if (flightStream.next()) {
+            vectorSchemaRoot = Optional.ofNullable(flightStream.getRoot());
+        }
+
+        if (!vectorSchemaRoot.isPresent()) {
+            completed = true;
+        }
+
+        if (isFinished()) {
+            return null;
+        }
+
+        currentPosition++;
+
+        List<Block> blocks = new ArrayList<>();
+        for (int columnIndex = 0; columnIndex < columnHandles.size(); columnIndex++) {
+            FieldVector vector = vectorSchemaRoot.get().getVector(columnIndex);
+            Type type = columnHandles.get(columnIndex).getColumnType();
+
+            boolean isDictionaryBlock = vector.getField().getDictionary() != null;
+            Dictionary dictionary = null;
+            if (isDictionaryBlock) {
+                dictionary = flightStream.getDictionaryProvider().lookup(vector.getField().getDictionary().getId());
+            }
+            Block block = null != dictionary ? ArrowPageUtils.buildBlockFromVector(vector, type, dictionary.getVector(), isDictionaryBlock) :
+                    ArrowPageUtils.buildBlockFromVector(vector, type, null, false);
```

Comment:
```suggestion
            Block block;
            if (vector.getField().getDictionary() != null) {
                Dictionary dictionary = flightStream.getDictionaryProvider().lookup(vector.getField().getDictionary().getId());
                block = buildBlockFromVector(vector, type, dictionary.getVector(), true);
            }
            else {
                block = buildBlockFromVector(vector, type, null, false);
            }
```

Reply:
Can we just always pass in a `DictionaryProvider` and check if it's a dictionary inside `buildBlockFromVector`?

Reply:
Refactored to pass in a DictionaryProvider to `buildBlockFromVector`